app/controllers/api/revoke_controller.rb

```diff
@@ -0,0 +1,50 @@
+module Api
+  class RevokeController < BaseController
+    def validate_method_action(_method_name, action_name)
+      super(:put, action_name)
+    end
+
+    def custom_api_user_role_allows_method?(_identifier)
+      true
+    end
+
+    def custom_api_user_role_allows?(action_identifier)
+      return true unless action_identifier
+
+      @cache_identifier ||= {}
+      @cache_identifier[action_identifier] ||= current_user.role_allows?(:identifier => action_identifier, :any => true)
+      @cache_identifier[action_identifier]
+    end
+
+    def filtered_user_from_params(userid)
+      user_object = User.find_by(:userid => userid)
+      raise BadRequestError, "Unable to find user #{userid}" unless user_object
+
+      Rbac::Filterer.filtered_object(user_object)
+    end
+
+    def user_from_params_allowed?(userid)
+      userid && userid != current_user.userid && current_user.role_allows?(:identifier => 'revoke_user_sessions')
+    end
+
+    def current_user_allowed?
+      current_user.role_allows?(:identifier => 'revoke_current_user_sessions')
+    end
+
+    def target_user
+      if user_from_params_allowed?(userid = params[:userid])
+        filtered_user_from_params(userid)
+      elsif current_user_allowed?
```

Comment:
I just want to make sure of the following:
- A user can revoke his own sessions with or without the userid param and without needing the revoke_user_sessions entitlement.
- A user can revoke someone else's session if and only if they have the revoke_user_sessions entitlement.

Reply:
I think it is not possible so simply as you suggesting and we need to use more entitlements.

- we need one entitlement(product feature added [here](https://github.com/ManageIQ/manageiq/pull/20633) 1- `revoke_session`) just to allow access method `update` -  this is done generally and this check is done in https://github.com/ManageIQ/manageiq-api/blob/master/app/controllers/api/base_controller/parser.rb#L178.
It finds entitlement in api.yml(added in this PR ) for edit action and then it is checked whether current user has this entitlement(product feature added)
- > - A user can revoke his own sessions with or without the userid param and without needing the revoke_user_sessions entitlement.

We are not able to revoke without `revoke_user_sessions ` but we need also  entitlement `revoke_session` to have on current user from first point and thanks to that we need to add new entitlement just for this case so I added `revoke_current_user_sessions`.

- > - A user can revoke someone else's session if and only if they have the revoke_user_sessions entitlement.

Yes this is correct, this operation can be performed only if current (authenticated) user has `revoke_user_sessions` entitlement(product feature) and if current (authenticated) user has RBAC access to this user.


Hopefully this is more clear and it makes sense now.
If you have any other concerns/suggestions please let me know.

thanks

Reply:
I wonder if you need to update https://github.com/ManageIQ/manageiq/blob/master/db/fixtures/miq_user_roles.yml  in the https://github.com/ManageIQ/manageiq/pull/20633 PR so you add the `revoke_user_sessions` to let's say admins and super-admins, and add `revoke_sessions` to every other role.

Reply:
yes that's a good note, I did update -
`super_administrator` and `administrator` 
have privileges(`revoke_sessions` as parent entitlement) to perform both revoke operations and
other roles have privileges(`revoke_current_user_sessions`) to perform **only** revoke own session of authenticated (current)user. 

thanks